pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,293 @@
+		excludedNamespaces = sets.NewString(params.Namespaces.Exclude...)
+	}
+
+	return thresholdPriority, includedNamespaces, excludedNamespaces, nil
+}
+
+func RemovePodsViolatingTopologySpreadConstraint(
+	ctx context.Context,
+	client clientset.Interface,
+	strategy api.DeschedulerStrategy,
+	nodes []*v1.Node,
+	podEvictor *evictions.PodEvictor,
+) {
+	thresholdPriority, includedNamespaces, excludedNamespaces, err := validateTopologySpreadParams(ctx, client, strategy.Params)
+	if err != nil {
+		klog.ErrorS(err, "Invalid PodLifeTime parameters")
+		return
+	}
+
+	nodeMap := make(map[string]*v1.Node, len(nodes))
+	for _, node := range nodes {
+		nodeMap[node.Name] = node
+	}
+	evictable := podEvictor.Evictable(evictions.WithPriorityThreshold(thresholdPriority))
+
+	// 1. for each namespace for which there is Topology Constraint
+	// 2. for each TopologySpreadyConstraint in that namespace
+	//  { find all evictable pods in that namespace
+	//  { 3. for each evictable pod in that namespace
+	// 4. If the pod matches this TopologySpreadConstraint LabelSelector
+	// 5. If the pod nodeName is present in the nodeMap
+	// 6. create a topoPair with key as this TopologySpreadConstraint.TopologyKey and value as this pod's Node Label Value for this TopologyKey
+	// 7. add the pod with key as this topoPair
+	// 8. find the min number of pods in any topoPair for this topologyKey
+	// iterate through all topoPairs for this topologyKey and diff currentPods -minPods <=maxSkew
+	// if diff > maxSkew, add this pod in the current bucket for eviction
+
+	// First record all of the constraints by namespace
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		if (!includedNamespaces.Has(namespace.Name) || excludedNamespaces.Has(namespace.Name)) && (includedNamespaces.Len()+excludedNamespaces.Len() > 0) {
+			continue
+		}
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "Couldn't list pods in namespace", "namespace", namespace)
+			return
+		}
+
+		// ...where there is a topology constraint
+		namespaceTopologySpreadConstrainPods := make([]v1.Pod, 0, len(namespacePods.Items))
+		for _, pod := range namespacePods.Items {
+			if len(pod.Spec.TopologySpreadConstraints) > 0 {
+				namespaceTopologySpreadConstrainPods = append(namespaceTopologySpreadConstrainPods, pod)
+			}
+		}
+		if len(namespaceTopologySpreadConstrainPods) == 0 {
+			return
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
+		for _, constrainedPod := range namespaceTopologySpreadConstrainPods {
+			// find pods to evict one constraint at a time
+			for _, constraint := range constrainedPod.Spec.TopologySpreadConstraints {
```

Comment:
So I think an ideal approach is to iterate over pods and gets a merged slice with tuple `{[]constraints, nodeSelector/nodeAffinity}` as its element. And later we process each tuple instead of each constraint.

Reply:
Good point, but I think we could also just cover this with a `map[v1.TopologySpreadConstraint]struct{}` (see new commit)